commit d06e15a7b03a2391b5bc38e458e118ab673b8c61
Author: Oliver Henshaw <oliver.henshaw@gmail.com>
Date:   Mon Jan 14 12:51:04 2013 +0000

    Stop grace period on unlock
    
    It was safe to leak the grace timer when it was in the lock process,
    since it was killed on unlock. Now it is not.
    
    A password should always be required if the screenlocker is explicitly
    triggered; but a screenlocker invoked while the grace timer from a
    previous invocation is still active can be dismissed without a password,
    at least for a short while. This also meant that the screenlocker
    keyboard shortcut would not work until the grace timer expired.
    
    REVIEW: 108425
    BUG: 313997
    FIXED-IN: 4.10.1

diff --git a/ksmserver/screenlocker/ksldapp.cpp b/ksmserver/screenlocker/ksldapp.cpp
index 10f808a..d940a95 100644
--- a/ksmserver/screenlocker/ksldapp.cpp
+++ b/ksmserver/screenlocker/ksldapp.cpp
@@ -243,6 +243,7 @@ void KSldApp::doUnlock()
     m_lockWindow = NULL;
     m_locked = false;
     m_lockedTimer.invalidate();
+    endGraceTime();
     KDisplayManager().setLock(false);
     emit unlocked();
     KNotification::event( QLatin1String("unlocked"));
@@ -331,6 +332,7 @@ bool KSldApp::isGraceTime() const
 
 void KSldApp::endGraceTime()
 {
+    m_graceTimer->stop();
     m_inGraceTime = false;
 }
 
